### PR TITLE
rio: add module

### PR DESCRIPTION
**Pull request by otavio**
### Description

Adds a `programs.rio` module to control Rio installation and configuration.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/programs/rio.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    # Only manage configuration if not empty
+    (lib.mkIf (cfg.settings != { } && !isDarwin) {
+      xdg.configFile."rio/config.toml".source = if lib.isPath cfg.settings then
+        cfg.settings
+      else
+        settingsFormat.generate "rio.toml" cfg.settings;
+    })
+
+    (lib.mkIf (cfg.settings != { } && isDarwin) {
+      home.file."Library/Application Support/rio/config.toml".source =
```
Comment (teto) [inline modules/programs/rio.nix:44]: I believe xdg.configFile does that automatically ?

File: modules/misc/news.nix [hunk 2]
```diff
@@ -1127,6 +1127,13 @@ in
           written in SVG. Supports many popular themes.
         '';
       }
+
+      {
+        time = "2023-06-19T16:18:30+00:00";
+        message = ''
+          A new module is available: 'programs.rio'.
```
Comment (teto) [inline modules/misc/news.nix:1134]: can you just mention what rio is/does to help users

File: modules/programs/rio.nix (hunk 1 above)
Comment (otavio) [reply modules/programs/rio.nix:44]: Are you sure? it doesn't seem to be the case. Can you point me the source code where this is set?

File: modules/programs/rio.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      description = lib.mdDoc ''
+        Enable Rio, a terminal built to run everywhere, as a native desktop applications by
+        Rust/WebGPU or even in the browsers powered by WebAssembly/WebGPU.
+      '';
+    };
+
+    package = lib.mkPackageOption pkgs "rio" { };
+
+    settings = lib.mkOption {
+      type = lib.types.either lib.types.path settingsFormat.type;
```
Comment (rycee) [inline modules/programs/rio.nix:20]: The settings option is intended only for structured configurations

```suggestion
      type = settingsFormat.type;
```

File: tests/modules/programs/rio/example-settings.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  else
+    ".config/rio/config.toml";
+
+  expected = pkgs.writeText "rio-expected.toml" ''
+    cursor = "_"
+    padding-x = 0
+    performance = "Low"
+  '';
+in {
+  config = {
```
Comment (rycee) [inline tests/modules/programs/rio/example-settings.nix:17]: Can unwrap the config attribute
```suggestion
```

File: tests/modules/programs/rio/empty-settings.nix [hunk 5]
```diff
@@ -0,0 +1,13 @@
+_:
+
+{
+  config = {
```
Comment (rycee) [inline tests/modules/programs/rio/empty-settings.nix:4]:
```suggestion
```

File: modules/programs/rio.nix (hunk 3 above)
Comment (rycee) [reply modules/programs/rio.nix:20]: If you want to allow the user to add arbitrary configuration, then you can add an `extraConfig` option of type `lines` that can be merged into the configuration file. But if the format really is plain toml then there should be no need, I believe the Nix syntax covers all toml features.

**Comment by Nyabinary** [comment]
What's the status on this so far? @otavio

**Comment by otavio** [comment]
AFAIK it is ready, I just need to rebase.

**Comment by Nyabinary** [comment]
@otavio Nice! I was going to make my own rio module but seeing as it's ready I'll just wait.

**Comment by otavio** [comment]
Please give it a try

**Comment by Nyabinary** [comment]
@otavio Did some light testing and it seems to work fine :)

**Comment by teto** [comment]
needs a rebase

File: modules/programs/rio.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      '';
+    };
+
+    package = lib.mkPackageOption pkgs "rio" { };
+
+    settings = lib.mkOption {
+      type = settingsFormat.type;
+      default = { };
+      description = ''
+        Configuration written to <filename>$XDG_CONFIG_HOME/rio/config.toml</filename>. See
```
[the remark this answers is not in the document]
Comment (otavio) [reply modules/programs/rio.nix]: Fixed.

**Pull request merged**